week3/homework/src/index.js

```diff
@@ -1,3 +1,119 @@
+  return new Promise((resolve, reject) => {
+    fs.readFile('todos.json', 'utf8', (err, data) => {
+      if (err) reject(err);
+      else resolve(JSON.parse(data));
+    });
+  });
+}
+
+function saveTodos(todos) {
+  return new Promise((resolve, reject) => {
+    fs.writeFile('todos.json', JSON.stringify(todos), 'utf8', err => {
+      if (err) reject(err);
+      else resolve();
+    });
+  });
+}
+
+function parseAndValidateTodo(request) {
+  const { todo } = request.body; // body is a string and we need to go to a json object
+  // we could here also put JSON.parse and also if we know every request comes will be json express gives us nicer way by => app.use(express.json()
+  if (todo == null) throw new Error('todo not set');
+  if (todo.description != null) todo.description = todo.description.trim();
+  if (todo.description == null || todo.description.length === 0)
+    throw new Error('Description not set');
+  return todo;
+}
+
+async function readTodos(req, res) {
+  let todos = await readTodosFromFile();
+  res.send(todos);
+}
+async function readTodo(req, res) {
+  let todos = await readTodosFromFile();
+  let todo = todos.find(todo => todo.id === req.params.id);
+  res.send(todo);
+}
+
+async function createTodo(req, res) {
+  let todo = parseAndValidateTodo(req);
+  let todos = await readTodosFromFile();
+  todo.id = uuidv4();
+  todos.push(todo);
+  await saveTodos(todos);
+  res.status = 201;
+  res.end();
+}
+
+async function updateTodo(req, res) {
+  let updateTodo = parseAndValidateTodo(req);
+  let todos = await readTodosFromFile();
+  let originalTodo = todos.find(todo => todo.id === req.params.id);
+  originalTodo.description = updateTodo.description;
+  await saveTodos(todos);
+  res.status = 200;
+  res.end();
+}
+async function markAsDone(req, res) {
+  let todos = await readTodosFromFile();
+  let todo = todos.find(todo => todo.id === req.params.id);
+  todo.done = true;
+  await saveTodos(todos);
+  res.status = 201;
+  res.end();
+}
+
+async function markAsNotDone(req, res) {
+  let todos = await readTodosFromFile();
+  let todo = todos.find(todo => todo.id === req.params.id);
+  todo.done = false;
+  await saveTodos(todos);
+  res.status = 200;
+  res.end();
+}
+
+// array of object  [{id: "asd", description: "todo1"}, {id: "yxc",description: "todo2"}]
+// id in the url and description in the body
+// there is nothing to parse
+async function deleteTodo(req, res) {
+  let todos = await readTodosFromFile();
+  let todo = todos.find(todo => todo.id === req.params.id);
+  let indexToDelete = todos.indexOf(todo);
+  todos.splice(indexToDelete, 1);
+  todo.push(todo);
+  await saveTodos(todos);
+  res.status = 200;
+  res.end();
+}
+async function clearTodos(req, res) {
+  let todos = await readTodosFromFile();
+  todos = [];
+  await saveTodos(todos);
+  res.status = 200;
+  res.end();
+}
+
+// this is methods
+app.get('/todos', (req, res) => readTodos(req, res));
+app.get('/todos/:id', (req, res) => readTodo(req, res));
```

Comment:
good to organize them by the verb and not in the way they are mentioned in the homework 👍